finat/enriched.py

```diff
@@ -0,0 +1,103 @@
+
+    def entity_dofs(self):
+        '''Return the map of topological entities to degrees of
+        freedom for the finite element.'''
+        from FIAT.mixed import concatenate_entity_dofs
+        return concatenate_entity_dofs(self.cell, self.elements)
+
+    def space_dimension(self):
+        '''Return the dimension of the finite element space.'''
+        return sum(elem.space_dimension() for elem in self.elements)
+
+    @cached_property
+    def index_shape(self):
+        return (self.space_dimension(),)
+
+    @cached_property
+    def value_shape(self):
+        '''A tuple indicating the shape of the element.'''
+        shape, = set(elem.value_shape for elem in self.elements)
+        return shape
+
+    def _compose_evaluations(self, results):
+        keys, = set(map(frozenset, results))
+
+        def merge(tables):
+            tables = tuple(tables)
+            zeta = self.get_value_indices()
+            tensors = []
+            for elem, table in zip(self.elements, tables):
+                beta_i = elem.get_indices()
+                tensors.append(gem.ComponentTensor(
+                    gem.Indexed(table, beta_i + zeta),
+                    beta_i
+                ))
+            beta = self.get_indices()
+            return gem.ComponentTensor(
+                gem.Indexed(gem.Concatenate(*tensors), beta),
+                beta + zeta
+            )
+        return {key: merge(result[key] for result in results)
+                for key in keys}
+
+    def basis_evaluation(self, order, ps, entity=None):
+        '''Return code for evaluating the element at known points on the
+        reference element.
+
+        :param order: return derivatives up to this order.
+        :param ps: the point set object.
+        :param entity: the cell entity on which to tabulate.
+        '''
+        results = [element.basis_evaluation(order, ps, entity)
+                   for element in self.elements]
+        return self._compose_evaluations(results)
+
+    def point_evaluation(self, order, refcoords, entity=None):
+        '''Return code for evaluating the element at an arbitrary points on
+        the reference element.
+
+        :param order: return derivatives up to this order.
+        :param refcoords: GEM expression representing the coordinates
+                          on the reference entity.  Its shape must be
+                          a vector with the correct dimension, its
+                          free indices are arbitrary.
+        :param entity: the cell entity on which to tabulate.
+        '''
+        results = [element.point_evaluation(order, refcoords, entity)
+                   for element in self.elements]
+        return self._compose_evaluations(results)
+
+
+def tree_map(f, *args):
+    """Like the built-in :py:func:`map`, but applies to a tuple tree."""
+    nonleaf, = set(isinstance(arg, tuple) for arg in args)
+    if nonleaf:
+        ndim, = set(map(len, args))
```

Comment:
I see, the latter is clearer I think (although I agree with the readability), and less likely to be removed at a later date.